client/utils.js

```diff
@@ -24,3 +24,47 @@ export function ce(elementName, className, child) {
 export function ctn(text) {
   return document.createTextNode(text);
 }
+
+const getSearch = (params) => {
+  let searches = [];
+
+  for (const key of Object.keys(params)) {
+    searches.push(`${ key }=${ params[key] || '' }`);
+  }
+
+  if (searches.length > 0) {
+    return `?${ searches.join('&') }`;
+  }
+
+  return '';
+};
+
+export class FtmUrl {
+  constructor(url) {
+    const parser = document.createElement('a');
+    parser.href = url && url.toString() || '';
+
+    const qs = parser.search.replace(/^\?/, '').split('&');
+    const searchparams = {};
+
+    for (const q of qs) {
+      const pair = q.split('=');
+      if (pair.length > 1 && pair[0]) {
+        searchparams[pair[0]] = pair[1] || '';
+      }
+    }
+
+    this.protocol = parser.protocol;
+    this.host = parser.host;
+    //this.hostname = parser.hostname;
+    //this.port = parser.port;
```

Comment:
these needed?

Reply:
They're currently not used. I originally had them in for completeness, but didn't want to think about managing conflicts if they were written to.